File: src/aoc2016/Day01.java
package aoc2016;

import java.awt.Point;
import java.io.IOException;
import java.nio.file.*;
import java.util.*;

public class Day01 {
  private int[][]    moves     = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };
  private Point[]    sol       = { new Point(), null };

  private Set<Point> visited   = new HashSet<>();
  private int        direction = 0;

  public void solve(String string) {
    for (String s : string.split(", ")) {
      direction = Math.floorMod(direction + (s.charAt(0) == 'L' ? -1 : 1), 4);

      int amount = Integer.parseInt(s.substring(1));
      while (amount-- > 0) {
        sol[0].translate(moves[direction][0], moves[direction][1]);

        if (sol[1] == null && visited.contains(sol[0]))
          sol[1] = new Point(sol[0]);
        else
          visited.add(sol[0]);
      }
    }
    System.out.println(String.format("Part one = %d\nPart two = %d", Math.abs(sol[0].x) + Math.abs(sol[0].y),
        Math.abs(sol[1].x) + Math.abs(sol[1].y)));
  }

  public static void main(String[] args) throws IOException {
    String s = new String(Files.readAllBytes(Paths.get("./input/2016/Day01_input.txt")));
    new Day01().solve(s);
  }
}

File: src/aoc2015/kot/Day01.kt
package aoc2015.kot

import java.io.File

object Day01 {
  fun partOne(input: String): Int = input.fold(0) { total, next -> total + if (next == '(') 1 else -1 }

  fun partTwo(input: String): Int {
    var pos = 0
    for ((idx, c) in input.withIndex()) {
      if (pos == -1) return idx
      when (c) {
        '(' -> pos++
        ')' -> pos--
      }
    }
    return -1
  }

}

fun main(args: Array<String>) {
  val input = File("./input/2015/Day01_input.txt").readText()
  println("Part One = ${Day01.partOne(input)}")
  println("Part Two = ${Day01.partTwo(input)}")
}
